Log in goroutine after ...

# Problem

Running the following tests from within `consensus/tendermint/core`

```
go test -run "TestStartRound|TestPrevoteTimeout|TestPrecommitTimeout" -count 100
```

Results in the following error.

```
panic: Log in goroutine after TestPrecommitTimeout/precommit_timeout_is_not_started_multiple_times has completed

goroutine 844 [running]:
testing.(*common).logDepth(0xc00030db00, 0xc00016e000, 0xe4, 0x3)
	/usr/local/go/src/testing/testing.go:680 +0x4d3
testing.(*common).log(...)
	/usr/local/go/src/testing/testing.go:662
testing.(*common).Fatalf(0xc00030db00, 0xb9f957, 0x2e, 0xc000178f00, 0x5, 0x5)
	/usr/local/go/src/testing/testing.go:723 +0x82
github.com/golang/mock/gomock.(*Controller).Call.func1(0xc00039ecf0, 0xb56840, 0xc000414060, 0xb7e351, 0x4, 0xc00078c940, 0x1, 0x1, 0x0, 0x0, ...)
	/home/pierspowlesland/go/pkg/mod/github.com/golang/mock@v1.4.3/gomock/controller.go:201 +0x485
github.com/golang/mock/gomock.(*Controller).Call(0xc00039ecf0, 0xb56840, 0xc000414060, 0xb7e351, 0x4, 0xc00078c940, 0x1, 0x1, 0xc00005ff70, 0xc000593420, ...)
	/home/pierspowlesland/go/pkg/mod/github.com/golang/mock@v1.4.3/gomock/controller.go:217 +0xb4
github.com/clearmatics/autonity/consensus/tendermint/core.(*MockBackend).Post(0xc000414060, 0xb05820, 0xc000593420)
	/home/pierspowlesland/projects/autonity/consensus/tendermint/core/backend_mock.go:182 +0xb7
github.com/clearmatics/autonity/consensus/tendermint/core.(*core).sendEvent(...)
	/home/pierspowlesland/projects/autonity/consensus/tendermint/core/handler.go:234
github.com/clearmatics/autonity/consensus/tendermint/core.(*core).onTimeoutPrecommit(0xc00039ca00, 0x0, 0xc000314940)
	/home/pierspowlesland/projects/autonity/consensus/tendermint/core/timeout.go:156 +0x17a
github.com/clearmatics/autonity/consensus/tendermint/core.(*timeout).scheduleTimeout.func1()
	/home/pierspowlesland/projects/autonity/consensus/tendermint/core/timeout.go:53 +0x37
created by time.goFunc
	/usr/local/go/src/time/sleep.go:168 +0x44
exit status 2
```

This error is raised by the go testing package here.

```
src/testing/testing.go
680:		panic("Log in goroutine after " + c.name + " has completed")
```

And it basically means we most likely called one of these functions `t.Log`, `t.Fatal` or `t.Fatalf` (there may be others but these are probably the ones we use most) in a goroutine that was started by the test and not shut down before the test shut down.

`t.Fatal` and `t.Fatalf` should only be called from within the main test goroutine so if these are being called in a different goroutine that indicates a problem with the test. If it is just `t.Log` that is being called then we just need to make sure we close the goroutine before we exit the test.

# Acceptance Criteria

Running the following from within `consensus/tendermint/core` should pass.

```
go test -run "TestStartRound|TestPrevoteTimeout|TestPrecommitTimeout" -count 100
```

--- a/consensus/tendermint/core/upon_condition_test.go
+++ b/consensus/tendermint/core/upon_condition_test.go
@@ -85,6 +85,10 @@ func TestStartRoundVariables(t *testing.T) {
 
 		// Check the initial consensus state
 		checkConsensusState(t, currentHeight, currentRound, propose, nil, int64(-1), nil, int64(-1), core)
+
+		// stop the timer to clean up
+		err := core.proposeTimeout.stopTimer()
+		assert.NoError(t, err)
 	})
 	t.Run("ensure round x state variables are updated correctly", func(t *testing.T) {
 		ctrl := gomock.NewController(t)
@@ -126,6 +130,10 @@ func TestStartRoundVariables(t *testing.T) {
 		core.startRound(context.Background(), currentRound+2)
 
 		checkConsensusState(t, currentHeight, currentRound+2, propose, currentBlock, currentRound, currentBlock2, currentRound+1, core)
+
+		// stop the timer to clean up
+		err := core.proposeTimeout.stopTimer()
+		assert.NoError(t, err)
 	})
 }
 
@@ -240,6 +248,10 @@ func TestStartRound(t *testing.T) {
 
 		assert.Equal(t, currentRound, core.Round())
 		assert.True(t, core.proposeTimeout.timerStarted())
+
+		// stop the timer to clean up
+		err := core.proposeTimeout.stopTimer()
+		assert.NoError(t, err)
 	})
 	t.Run("at proposal timeout expiry timeout event is sent", func(t *testing.T) {
 		currentHeight := big.NewInt(int64(rand.Intn(maxSize) + 1))
@@ -593,6 +605,10 @@ func TestPrevoteTimeout(t *testing.T) {
 		err := c.handleCheckedMsg(context.Background(), prevoteMsg, members[sender])
 		assert.Nil(t, err)
 		assert.True(t, c.prevoteTimeout.timerStarted())
+
+		// stop the timer to clean up
+		err = c.prevoteTimeout.stopTimer()
+		assert.NoError(t, err)
 	})
 	t.Run("prevote timeout is not started multiple times", func(t *testing.T) {
 		currentHeight := big.NewInt(int64(rand.Intn(maxSize) + 1))
@@ -630,6 +646,9 @@ func TestPrevoteTimeout(t *testing.T) {
 		assert.True(t, c.prevoteTimeout.timerStarted())
 		assert.True(t, c.prevoteTimeout.start.Before(timeNow))
 
+		// stop the timer to clean up
+		err = c.prevoteTimeout.stopTimer()
+		assert.NoError(t, err)
 	})
 	t.Run("at prevote timeout expiry timeout event is sent", func(t *testing.T) {
 		currentHeight := big.NewInt(int64(rand.Intn(maxSize) + 1))
@@ -873,6 +892,10 @@ func TestPrecommitTimeout(t *testing.T) {
 		err := c.handleCheckedMsg(context.Background(), precommitMsg, members[sender])
 		assert.Nil(t, err)
 		assert.True(t, c.precommitTimeout.timerStarted())
+
+		// stop the timer to clean up
+		err = c.precommitTimeout.stopTimer()
+		assert.NoError(t, err)
 	})
 	t.Run("precommit timeout is not started multiple times", func(t *testing.T) {
 		currentHeight := big.NewInt(int64(rand.Intn(maxSize) + 1))
@@ -911,6 +934,9 @@ func TestPrecommitTimeout(t *testing.T) {
 		assert.True(t, c.precommitTimeout.timerStarted())
 		assert.True(t, c.precommitTimeout.start.Before(timeNow))
 
+		// stop the timer to clean up
+		err = c.precommitTimeout.stopTimer()
+		assert.NoError(t, err)
 	})
 	t.Run("at precommit timeout expiry timeout event is sent", func(t *testing.T) {
 		currentHeight := big.NewInt(int64(rand.Intn(maxSize) + 1))
@@ -961,6 +987,10 @@ func TestPrecommitTimeout(t *testing.T) {
 
 		assert.Equal(t, currentRound+1, c.Round())
 		assert.Equal(t, propose, c.step)
+
+		// stop the timer to clean up, since start round can start propose timeout
+		err := c.proposeTimeout.stopTimer()
+		assert.NoError(t, err)
 	})
 }
 

